### PR TITLE
Fix #12824

### LINKED ISSUE #12824: OpenCL: some kernels are broken for CV_16F
Build example: [1](http://pullrequest.opencv.org/buildbot/builders/master-lin64/builds/10608/steps/test_core-ippicv-opencl/logs/stdio)

Message:
```
[ RUN      ] OCL_MatrixOperation/UMatExpr.Eye/28, where GetParam() = (CV_16F, Channels(1))
OpenCL program build log: core/set_identity
Status -11: CL_BUILD_PROGRAM_FAILURE
-D T=? -D T1=? -D cn=1 -D ST=? -D kercn=1 -D rowsPerWI=4 -D INTEL_DEVICE
2:11:1: error: expected parameter declarator
```

### PR DESCRIPTION
**Pull request by dkurt**
### This pullrequest changes

resolves https://github.com/opencv/opencv/issues/12824 for GPUs with `cl_khr_fp16` features.

### CODE REVIEW

File: modules/core/src/ocl.cpp [hunk 1]
```diff
@@ -6002,10 +6002,10 @@ const char* typeToStr(int type)
         "int", "int2", "int3", "int4", 0, 0, 0, "int8", 0, 0, 0, 0, 0, 0, 0, "int16",
         "float", "float2", "float3", "float4", 0, 0, 0, "float8", 0, 0, 0, 0, 0, 0, 0, "float16",
         "double", "double2", "double3", "double4", 0, 0, 0, "double8", 0, 0, 0, 0, 0, 0, 0, "double16",
+        "half", "half2", "half3", "half4", 0, 0, 0, "half8", 0, 0, 0, 0, 0, 0, 0, "half16",
         "?", "?", "?", "?", "?", "?", "?", "?", "?", "?", "?", "?", "?", "?", "?", "?"
```
[the remark this answers is not in the document]
Comment (dkurt) [reply modules/core/src/ocl.cpp]: Replaced zeros to "?".

File: modules/core/src/ocl.cpp [hunk 2]
```diff
@@ -6038,10 +6038,10 @@ const char* vecopTypeToStr(int type)
         "int", "int2", "int3", "int4", 0, 0, 0, "int8", 0, 0, 0, 0, 0, 0, 0, "int16",
         "int", "int2", "int3", "int4", 0, 0, 0, "int8", 0, 0, 0, 0, 0, 0, 0, "int16",
         "ulong", "ulong2", "ulong3", "ulong4", 0, 0, 0, "ulong8", 0, 0, 0, 0, 0, 0, 0, "ulong16",
+        "half", "half2", "half3", "half4", 0, 0, 0, "half8", 0, 0, 0, 0, 0, 0, 0, "half16",
```
[the remark this answers is not in the document]
Comment (dkurt) [reply modules/core/src/ocl.cpp]: Done, thank you!

File: modules/core/src/ocl.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (alalek) [reply modules/core/src/ocl.cpp]: I believe it is better to `CV_Assert(result != NULL); return result;` (and completely remove useless "?")

File: modules/core/src/opencl/set_identity.cl [hunk 3]
```diff
@@ -43,6 +43,10 @@
 //
 //M*/
 
+#if defined(cl_khr_fp16)
+#pragma OPENCL EXTENSION cl_khr_fp16 : enable
+#endif
```
[the remark this answers is not in the document]
Comment (dkurt) [reply modules/core/src/opencl/set_identity.cl]: @alalek, May I ask you to specify which type should be replaced?

[the remark this answers is not in the document]
Comment (alalek) [reply modules/core/src/opencl/set_identity.cl]: This kernel has "copy/fill" type. It is not necessary to use "actual type" here directly.
For example, "float" type is replaced by "int". "half" should be replaced by "short".

So after fixing `memopTypeToStr()`, enabling of this extension is not required.

**Pull request merged**